File: packages/python/envguard/validator.py
"""EnvGuard validation API."""

import json
import subprocess
from dataclasses import dataclass
from typing import List, Optional, Union

from .install import ensure_binary


@dataclass
class ValidationError:
    key: str
    message: str
    rule: str


@dataclass
class ValidationResult:
    valid: bool
    errors: List[ValidationError]
    warnings: List[ValidationError]


def validate(
    schema_path: Optional[str] = None,
    env_path: Optional[Union[str, List[str]]] = None,
    strict: bool = False,
    env_name: Optional[str] = None,
) -> ValidationResult:
    """Validate a .env file against a schema.

    Args:
        schema_path: Path to the schema YAML file. Defaults to "envguard.yaml".
        env_path: Path to the .env file, or list of paths. Defaults to ".env".
        strict: Fail if .env contains keys not defined in schema.
        env_name: Environment name (e.g. "production", "development") for environment-specific rules.

    Returns:
        ValidationResult with valid flag, errors, and warnings.

    Raises:
        RuntimeError: If EnvGuard binary fails or returns unexpected output.
    """
    binary = ensure_binary()
    args = [str(binary), "validate", "--format", "json"]

    if schema_path:
        args.extend(["--schema", schema_path])
    if env_path:
        if isinstance(env_path, str):
            env_path = [env_path]
        for p in env_path:
            args.extend(["--env", p])
    if strict:
        args.append("--strict")
    if env_name:
        args.extend(["--env-name", env_name])

    result = subprocess.run(
        args,
        capture_output=True,
        text=True,
    )

    # Try to parse JSON from stdout
    stdout = result.stdout.strip()
    if stdout:
        try:
            data = json.loads(stdout)
            return ValidationResult(
                valid=data.get("valid", False),
                errors=[ValidationError(**e) for e in data.get("errors", [])],
                warnings=[ValidationError(**w) for w in data.get("warnings", [])],
            )
        except (json.JSONDecodeError, TypeError):
            pass

    # If we can't parse JSON, raise an error
    stderr = result.stderr.strip()
    raise RuntimeError(
        f"EnvGuard failed (exit code {result.returncode}): "
        f"{stderr or stdout or 'Unknown error'}"
    )

File: packages/node/src/validator.ts
import { spawn } from "child_process";
import * as path from "path";
import type { ValidateOptions, ValidationResult } from "./types";
import { getBinaryDir, getBinaryName } from "./install";

function getBinaryPath(): string {
  return path.join(getBinaryDir(), getBinaryName());
}

export function validate(options: ValidateOptions = {}): Promise<ValidationResult> {
  const binaryPath = getBinaryPath();
  const args = ["validate", "--format", "json"];

  if (options.schemaPath) {
    args.push("--schema", options.schemaPath);
  }
  if (options.envPath) {
    args.push("--env", options.envPath);
  }
  if (options.strict) {
    args.push("--strict");
  }
  if (options.envName) {
    args.push("--env-name", options.envName);
  }

  return new Promise((resolve, reject) => {
    const proc = spawn(binaryPath, args, {
      stdio: ["ignore", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";

    proc.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    proc.stderr.on("data", (data) => {
      stderr += data.toString();
    });

    proc.on("close", (code) => {
      // Try to parse JSON from stdout regardless of exit code
      const jsonMatch = stdout.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        try {
          const result = JSON.parse(jsonMatch[0]) as ValidationResult;
          resolve(result);
          return;
        } catch {
          // Fall through to reject
        }
      }

      if (code === 2) {
        reject(new Error(`EnvGuard error: ${stderr || stdout}`));
      } else if (code === 1) {
        // Validation failed but we couldn't parse JSON
        reject(new Error(`Validation failed: ${stderr || stdout}`));
      } else {
        reject(new Error(`Unexpected exit code ${code}: ${stderr || stdout}`));
      }
    });

    proc.on("error", (err) => {
      reject(new Error(`Failed to run EnvGuard: ${err.message}`));
    });
  });
}

export function validateSync(options: ValidateOptions = {}): ValidationResult {
  const { execFileSync } = require("child_process");
  const binaryPath = getBinaryPath();
  const args = ["validate", "--format", "json"];

  if (options.schemaPath) {
    args.push("--schema", options.schemaPath);
  }
  if (options.envPath) {
    args.push("--env", options.envPath);
  }
  if (options.strict) {
    args.push("--strict");
  }
  if (options.envName) {
    args.push("--env-name", options.envName);
  }

  try {
    const stdout = execFileSync(binaryPath, args, { encoding: "utf-8" });
    return JSON.parse(stdout) as ValidationResult;
  } catch (err: any) {
    if (err.stdout) {
      try {
        return JSON.parse(err.stdout.toString()) as ValidationResult;
      } catch {
        // Fall through
      }
    }
    throw new Error(`EnvGuard failed: ${err.message}`);
  }
}
